src/combinatorics/simple_graph/prod.lean

```diff
@@ -0,0 +1,154 @@
+/-
+Copyright (c) 2022 George Peter Banyard, Yaël Dillies, Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: George Peter Banyard, Yaël Dillies, Kyle Miller
+-/
+import combinatorics.simple_graph.connectivity
+
+/-!
+# Graph products
+
+This file defines the box product of graphs and other product constructions. The box product of `G`
+and `H` is the graph on the product of the vertices such that `x` and `y` are related iff they agree
+on one component and the other one is related via either `G` or `H`. For example, the box product of
+two edges is a square.
+
+## Main declarations
+
+* `simple_graph.box_prod`: The box product.
+
+## Notation
+
+* `G □ H`: The box product of `G` and `H`.
+
+## TODO
+
+Define all other graph products!
+-/
+
+variables {α β γ : Type*}
+
+namespace simple_graph
+variables {G : simple_graph α} {H : simple_graph β} {I : simple_graph γ} {a a₁ a₂ : α} {b b₁ b₂ : β}
+  {x y : α × β}
+
+/-- Box product of simple graphs. It relates `(a₁, b)` and `(a₂, b)` if `G` relates `a₁` and `a₂`,
+and `(a, b₁)` and `(a, b₂)` if `H` relates `b₁` and `b₂`. -/
+def box_prod (G : simple_graph α) (H : simple_graph β) : simple_graph (α × β) :=
+{ adj := λ x y, G.adj x.1 y.1 ∧ x.2 = y.2 ∨ H.adj x.2 y.2 ∧ x.1 = y.1,
+  symm := λ x y, by simp [and_comm, or_comm, eq_comm, adj_comm],
+  loopless := λ x, by simp }
+
+infix ` □ `:70 := box_prod
+
+lemma box_prod_adj : (G □ H).adj x y ↔ G.adj x.1 y.1 ∧ x.2 = y.2 ∨ H.adj x.2 y.2 ∧ x.1 = y.1 :=
+iff.rfl
+
+@[simp] lemma box_prod_adj_left : (G □ H).adj (a₁, b) (a₂, b) ↔ G.adj a₁ a₂ :=
+by rw [box_prod_adj, and_iff_left rfl, or_iff_left (λ h : H.adj b b ∧ _, h.1.ne rfl)]
+
+@[simp] lemma box_prod_adj_right : (G □ H).adj (a, b₁) (a, b₂) ↔ H.adj b₁ b₂ :=
+by rw [box_prod_adj, and_iff_left rfl, or_iff_right (λ h : G.adj a a ∧ _, h.1.ne rfl)]
+
+variables (G H I)
+
+/-- The box product is commutative up to isomorphism. `equiv.prod_comm` as a graph isomorphism. -/
+@[simps] def box_prod_comm : G □ H ≃g H □ G := ⟨equiv.prod_comm _ _, λ x y, or_comm _ _⟩
+
+/-- The box product is associative up to isomorphism. `equiv.prod_assoc` as a graph isomorphism. -/
+@[simps] def box_prod_assoc : (G □ H) □ I ≃g G □ (H □ I) :=
+⟨equiv.prod_assoc _ _ _, λ x y, by simp only [box_prod_adj, equiv.prod_assoc_apply,
+  or_and_distrib_right, or_assoc, prod.ext_iff, and_assoc, @and.comm (x.1.1 = _)]⟩
+
+namespace walk
+variables {G}
+
+/-- Turn a walk on `G` into a walk on `G □ H`. -/
+def box_prod_left (b : β) : Π {a₁ a₂ : α}, G.walk a₁ a₂ → (G □ H).walk (a₁, b) (a₂, b)
+| _ _ nil := nil
+| _ _ (cons h w) := w.box_prod_left.cons $ box_prod_adj_left.2 h
```

Comment:
I think this would be better using `simple_graph.walk.map` after defining inclusion homomorphisms `G ->g G □ H` for each `b : β` (or, rather, embeddings, right?)